### ISSUE TITLE
Add a GC root for home-manager

### ISSUE THREAD

**Issue by sgraf812** [rycee/home-manager#305]
Currently, everytime I `nix-collect-garbage`, my `home-manager` profile gets collected, too. Is this intended? If so, what would I do to collect old generations?

**Comment by rycee** [comment]
No, that definitely is not intended. Which output do you get if you run the following command?

```bash
ls --sort=time -l /nix/var/nix/gcroots/per-user/$(whoami) /nix/var/nix/profiles/per-user/$(whoami)/home-manager*
```

**Comment by rycee** [comment]
Sorry for the delay. You're having a curious problem that I've never seen before. Everything seems as it should be in the output of the `ls` command.

Unfortunately I'm not certain what it could be caused by. I'll just dump a bunch of questions and hopefully something will lead further.

So after a garbage collection, links that are created by Home Manager in your home directory end up not pointing to anything? What about your `~/.nix-profile` directory, is it completely emptied?

Which precise `nix-collect-garbage` command are you running?

If you run `nix-collect-garbage`, then `/nix/var/nix/gcroots/per-user/sgraf/current-home/activate`, and finally `nix-collect-garbage -v` does the second garbage collection print the things HM just created?

Does `nix-store --gc --print-roots | grep home-manager` print similar paths as the `ls` command?

Are you using Home Manager to manage your Bash or Z shell? If so do you have
```shell
. .nix-profile/etc/profile.d/nix.sh
```
in your `~/.profile`?

**Comment by sgraf812** [comment]
Actually, I hit this today in a less grave variant: I ran `nix-collect-garbage -d` (emphasis on -d) and it removed all home-manager and nix tools again, at least from my path. Maybe it's related to me running `nix store --optmize` before. Also, re-installing wasn't so involved this time, because it seems that my actual home-manager generation was still intact, so it just re-downloaded 10ish packages and I was good to go again in no time.

**Comment by rycee** [comment]
Sorry for the delay. If these packages are installed in your latest Home Manager profile then you shouldn't have to download them again. You can check the output of `home-manager packages` and the actual content of the HM profile path to see if the packages are part of the HM profile. For example, for the home-manager tool I have
```console
$ home-manager packages | grep '^home-manager$'
home-manager
$ find $(nix-env -q --out-path | grep -o '/.*home-manager-path') -path '*/bin/home-manager'
/nix/store/sr67x7l0n41lfbrhn5kb86ann07rkazx-home-manager-path/bin/home-manager
```

**Comment by rycee** [comment]
Yeah, these packages are needed to build the actual Home Manager generation but are not needed afterwards so the GC will remove them.
